closed/src/java.base/share/classes/openj9/internal/criu/CRIUSECProvider.java

```diff
@@ -51,6 +51,9 @@ public CRIUSECProvider() {
 
         // SHA1PRNG is the default name needed by the jdk, but SHA1 is not used, rather it reads directly from /dev/random.
         putService(new Service(this, "MessageDigest", "SHA-1", packageName + "SHA", java.util.Arrays.asList(aliases), null));
+        putService(new Service(this, "MessageDigest", "SHA-256", packageName + "SHA2$SHA256", null, null));
+        putService(new Service(this, "MessageDigest", "MD5", packageName + "MD5", null, null));
+        putService(new Service(this, "Mac", "HmacSHA256", packageName + "HmacCore$HmacSHA256", null, null));
```

Comment:
Why can't we simply use the classes in their existing packages (as opposed to making copies)?
Why should we add MD5, a hash function widely considered broken?

Reply:
`Why can't we simply use the classes in their existing packages (as opposed to making copies)?`
The reason why we did not simply use the classes in the existing packages is because MD5 and SHA256 need to inherit from a DigestBase class. For CRIU, we have a custum CRIU DigestBase class file, and there is a reset function defined in it such that we can reset the digest to ensure no state during checkpoint. But in the existing packages, there is no such resetDigest method implementation in other DigestBase class. e.g. in the SUN provider, https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/415481a396eeb9df3ae4655447fa8998dfe40a2d/src/java.base/share/classes/sun/security/provider/DigestBase.java. We just want to follow the existing patterns within the CRIU folder.

`Why should we add MD5, a hash function widely considered broken?`
we know MD5 is antiquated and not suggested for the most secure hashing use cases. But the caller in this case liberty making use of spring boot needed to use it.

Reply:
The list of copies of upstream code is getting unmanageable. The existing copies appear to have fallen behind and may be missing CVE fixes.

Instead of copying more code, I suggest we move the list from `DigestBase` to `SecurityProviders`. The new list would contain `Runnable` objects, each of which would be `run()` where the current call to `CRIUSECProvider.resetCRIUSEC()` is made.

Instead of adding objects to `DigestBase.digestList`, the constructor of `DigestBase` would create a reset `Runnable` and provide it to a method of `SecurityProviders` for tracking. The changes to `DigestBase` can be conditional on `CRIU_SUPPORT`.  `NativePRNG` can be changed in a similar manner rather than duplicating the code.

Reply:
Sounds good, let's me update the code.

Reply:
It also occurs to me that `DigestBase.digestList` represents a leak (objects are never removed from that list); we'll need a `SoftReference` somewhere in the implementation to prevent that list from growing without bound. Objects that are otherwise unreferenced need not be reset.